Documentation/Evolution/RegexSyntax.md

```diff
@@ -0,0 +1,871 @@
+- `UTF`: Enables UTF pattern support.
+- `UCP`: Enables Unicode property support.
+
+### Callouts
+
+```
+Callout -> PCRECallout | OnigurumaCallout
+
+PCRECallout -> '(?C' CalloutBody ')'
+PCRECalloutBody -> '' | <Number>
+                 | '`' <String> '`'
+                 | "'" <String> "'"
+                 | '"' <String> '"'
+                 | '^' <String> '^'
+                 | '%' <String> '%'
+                 | '#' <String> '#'
+                 | '$' <String> '$'
+                 | '{' <String> '}'
+                 
+OnigurumaCallout -> OnigurumaNamedCallout | OnigurumaCalloutOfContents
+
+OnigurumaNamedCallout   -> '(*' Identifier OnigurumaTag? OnigurumaCalloutArgs? ')'
+OnigurumaCalloutArgs    -> '{' OnigurumaCalloutArgList '}'
+OnigurumaCalloutArgList -> OnigurumaCalloutArg (',' OnigurumaCalloutArgList)*
+OnigurumaCalloutArg     -> [^,}]+
+OnigurumaTag            -> '[' Identifier ']'
+
+OnigurumaCalloutOfContents -> '(?' '{'+ Contents '}'+ OnigurumaTag? Direction? ')'
+OnigurumaCalloutContents   -> <String>
+OnigurumaCalloutDirection  -> 'X' | '<' | '>'
+```
+
+A callout is a feature that allows a user-supplied function to be called when matching reaches that point in the pattern. We supported parsing both the PCRE and Oniguruma callout syntax. The PCRE syntax accepts a string or numeric argument that is passed to the function. The Oniguruma syntax is more involved, and may accept a tag, argument list, or even an arbitrary program in the 'callout of contents' syntax.
+
+### Absent functions
+
+```
+AbsentFunction -> '(?~' RegexNode ')'
+                | '(?~|' Concatenation '|' Concatenation ')'
+                | '(?~|' Concatenation ')'
+                | '(?~|)'
+```
+
+An absent function is an Oniguruma feature that allows for the easy inversion of a given pattern. There are 4 variants of the syntax:
+
+- `(?~|absent|expr)`: Absent expression, which attempts to match against `expr`, but is limited by the range that is not matched by `absent`.
+- `(?~absent)`: Absent repeater, which matches against any input not matched by `absent`. Equivalent to `(?~|absent|\O*)`.
+- `(?~|absent)`: Absent stopper, which limits any subsequent matching to not include `absent`.
+- `(?~|)`: Absent clearer, which undoes the effects of the absent stopper.
+
+## Syntactic differences between engines
+
+The above regular expression grammar covers a superset of the syntax accepted by PCRE, ICU, Oniguruma, .NET, and Java. However there are cases where the same syntax is parsed differently by these engines. This section provides a summary of these differences, and specifies the interpretation that will be made by the Swift regex parser.
+
+These differences inherently mean that our default parser behavior cannot be fully compatible with these other engines. However, this would not preclude the potential future implementation of different compatibility modes for different engines in which we support their parsing behavior of certain syntax.
+
```

Comment:
Huh, I thought this was marked for me to fill out, so I did. But now I don't see the notes. Here's what I wrote in case it's helpful, not that you have to take it.

> The proposed "syntactic superset" introduces some minor ambiguities, as each engine supports a slightly different set of features. When a particular engine's parser sees a feature it doesn't support, it typically has a fall-back behavior, such as treating the unknown feature as literal contents.
>
>Explicit compatibility modes, i.e. precisely mimicing emergent behavior from a specific engine's parser, is deferred as future work from this proposal. Conversion from this "syntactic superset" to a particular engine's syntax (e.g. as an AST "pretty printer") is deferred as future work from this proposal.
>
>Below is an exhaustive treatment of every syntactic ambiguity we have encountered.

Reply:
I kind of prefer this version, updated to take it